src/integration/resources/inprocess/dbnode.go

```diff
@@ -0,0 +1,329 @@
+	// NOTES:
+	// - id we can generate
+	// - isolation_group set a constant?
+	// - weight just use 1024?
+	// - address is 0.0.0.0
+	// - port is the listen address (get from config)
+
+	return &admin.Host{
+		Id:             "foo",
+		IsolationGroup: "foo-a",
+		Zone:           "embedded",
+		Weight:         1024,
+		Address:        "0.0.0.0",
+		Port:           uint32(port),
+	}, nil
+}
+
+func (d *dbNode) Health() (*rpc.NodeHealthResult_, error) {
+	return d.tchanClient.TChannelClientHealth(defaultRPCTimeout)
+}
+
+func (d *dbNode) WaitForBootstrap() error {
+	return retry(func() error {
+		health, err := d.Health()
+		if err != nil {
+			return err
+		}
+
+		if !health.GetBootstrapped() {
+			err = fmt.Errorf("not bootstrapped")
+			d.logger.Error("node not bootstrapped", zap.Error(err))
+			return err
+		}
+
+		return nil
+	})
+}
+
+func (d *dbNode) WritePoint(req *rpc.WriteRequest) error {
+	return d.tchanClient.TChannelClientWrite(defaultRPCTimeout, req)
+}
+
+func (d *dbNode) WriteTaggedPoint(req *rpc.WriteTaggedRequest) error {
+	return d.tchanClient.TChannelClientWriteTagged(defaultRPCTimeout, req)
+}
+
+func (d *dbNode) AggregateTiles(req *rpc.AggregateTilesRequest) (int64, error) {
+	res, err := d.tchanClient.TChannelClientAggregateTiles(defaultRPCTimeout, req)
+	if err != nil {
+		return 0, err
+	}
+
+	return res.ProcessedTileCount, nil
+}
+
+func (d *dbNode) Fetch(req *rpc.FetchRequest) (*rpc.FetchResult_, error) {
+	return d.tchanClient.TChannelClientFetch(defaultRPCTimeout, req)
+}
+
+func (d *dbNode) FetchTagged(req *rpc.FetchTaggedRequest) (*rpc.FetchTaggedResult_, error) {
+	return d.tchanClient.TChannelClientFetchTagged(defaultRPCTimeout, req)
+}
+
+func (d *dbNode) Exec(commands ...string) (string, error) {
+	//nolint:gosec
+	cmd := exec.Command(commands[0], commands[1:]...)
+
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	if err := cmd.Run(); err != nil {
+		return "", err
+	}
+
+	return out.String(), nil
+}
+
+func (d *dbNode) GoalStateExec(verifier resources.GoalStateVerifier, commands ...string) error {
+	return retry(func() error {
+		if err := verifier(d.Exec(commands...)); err != nil {
+			d.logger.Info("goal state verification failed. retrying")
+			return err
+		}
+		return nil
+	})
+}
+
+func (d *dbNode) Restart() error {
+	if err := d.Close(); err != nil {
+		return err
+	}
+
+	d.start()
+
+	return nil
+}
+
+func (d *dbNode) Close() error {
+	defer func() {
+		for _, dir := range d.tmpDirs {
+			if err := os.RemoveAll(dir); err != nil {
+				d.logger.Error("error removing temp directory", zap.String("dir", dir), zap.Error(err))
+			}
+		}
+	}()
+
+	for i := 0; i < d.cfg.Components(); i++ {
+		select {
+		case d.interruptCh <- errors.New("in-process node being shut down"):
+		case <-time.After(interruptTimeout):
+			return errors.New("timeout sending interrupt. closing without graceful shutdown")
+		}
+	}
+
+	for i := 0; i < d.cfg.Components(); i++ {
+		select {
+		case <-d.shutdownCh:
```

Comment:
What exactly are we looping over here? And what are pulling from the `shutdownCh`?

Reply:
A dbnode can be just a dbnode or a dbnode and an embedded coordinator. Therefore, we need to receive shutdown signals from however many components have been started.

Reply:
Yeah I am just confused by what we're looping over here? We don't seem to grab anything from the components?

Reply:
We're waiting on N many shutdown notifications before we proceed where N is equal to the number of components we started up, which is what `d.cfg.Components()` returns.

Reply:
Got it. So I think it would just make sense to make this a method for the server interface...

```
server.ShutdownComponents(waitTime time.Duration) error
```

with each component implementing an actual `Shutdown()` method

```
errors := make(chan error)
for _, c : range components {
  c := c
  go func() {
   // capture the error here
   errors <- c.Shutdown()
  } ()
}

// Wait for shutdowns, with appropriate the timeout and error catching.
```

The two channel setup is just feels awkward and hard to read.

Reply:
Yeah, I think moving this to the server package is a good call. Though, fwiw, there are still going to be two channels -- one to interrupt the components and one to get notified when the components are done shutting down. Will try to make this as friendly as possible